### PR TITLE
Rewrite ALCT trigger patterns as 2D vector (ACLUT-2)

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

Similar PR as #https://github.com/cms-sw/cmssw/pull/28846. ALCT patterns are rewritten as 2D vectors instead of 1D arrays. This allows us to implement an algorithm similar to CCLUT #30103 in the ALCT processor for Phase-2 studies. As this PR merely rewrites the patterns in a more convenient format, there should no changes in any of the workflows.

Note: `code-format` rewrote the [ALCT patterns](https://github.com/cms-sw/cmssw/compare/master...dildick:from-CMSSW_11_2_X_2020-06-02-2300-recast-alct-patterns?expand=1#diff-a7b633e207161fd21666160aff6ed802) as a single line instead. Originally I wrote

```
    // Collision pattern A                                                                                                                                  
    {{1, 1, 1, 0, 0},
     {0, 1, 1, 0, 0},
     {0, 0, 1, 0, 0},
     {0, 0, 1, 1, 0},
     {0, 0, 1, 1, 1},
     {0, 0, 1, 1, 1}},
```

#### PR validation:

Tested with WF 27434.0.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

@tahuang1991

### CODE REVIEW

**Comment by dildick** [comment]
@rekovic I'll do a data vs emulator comparison on the ALCTs with Run-2 2018 data, to make sure that the 2D vectors work as expected.

**Comment by dildick** [comment]
comparison_alct_bx_mAll
comparison_alct_bx_pAll
comparison_alct_keywire_mAll
comparison_alct_keywire_pAll
comparison_alct_quality_mAll
comparison_alct_quality_pAll
comparison_alct_valid_mAll
comparison_alct_valid_pAll

**Comment by dildick** [comment]
@rekovic Can you run the tests, please?

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
Do you have any comments @rekovic ?

File: L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h [hunk 1]
```diff
@@ -10,26 +10,21 @@
 
 class CSCPatternBank {
 public:
-  typedef std::vector<std::vector<std::vector<int> > > CLCTPatterns;
-
-  typedef int ALCTPatterns[CSCConstants::NUM_ALCT_PATTERNS][CSCConstants::MAX_WIRES_IN_PATTERN];
+  // typedef used for both ALCT and CLCT
+  typedef std::vector<std::vector<std::vector<int> > > LCTPatterns;
```
Comment (davidlange6) [inline L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h:14]: this is better done with std::array

Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h:14]: Perhaps, in that case I need different typedefs for ALCT and CLCT.

Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/interface/CSCPatternBank.h:14]: A construct like ```typedef std::vector<std::array<std::array<int, CSCConstants::ALCT_PATTERN_WIDTH>, CSCConstants::NUM_LAYERS > > ALCTPatterns;``` might be okay for ALCT, but ```typedef std::vector<std::array<std::array<int, CSCConstants::CLCT_PATTERN_WIDTH + 2>, CSCConstants::NUM_LAYERS > > CLCTPatterns;``` is too wide layers 1-5 for CLCT. ```typedef std::vector<std::array<std::vector<int>, CSCConstants::NUM_LAYERS > > LCTPatterns;``` would work for both without changing the patterns.

**Comment by dildick** [comment]
@rekovic Do you have any further comments?

**Comment by silviodonato** [comment]
@rekovic, can you please check and eventually sign?

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**